sklearn/metric_learning/nca.py

```diff
@@ -0,0 +1,280 @@
+from ..base import BaseEstimator, TransformerMixin
+from ..utils.validation import check_is_fitted
+
+import numpy as np
+import scipy as sp
+import scipy.misc
+import scipy.optimize
+import time
+
+import sys
+
+
+def nca_vectorized_oracle(X, y, n_components, loss, threshold=0.0):
+    dx = X[:, np.newaxis, :] - X[np.newaxis, :, :]  # n_samples x n_samples x n_features
+    outer = dx[:, :, np.newaxis, :] * dx[:, :, :, np.newaxis]  # n_samples x n_samples x n_features x n_features
```

Comment:
(and yes, caching those outer products would be out of the window, but we don't know yet how useful the caching actually is, since that tensor is incredibly redundant due to its low rank. This is the point I would like to attack be evaluating selectively, sequentially.)